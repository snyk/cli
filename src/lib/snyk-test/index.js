module.exports = test;

const detect = require('../detect');
const { runTest } = require('./run-test');
const chalk = require('chalk');
const pm = require('../package-managers');
const { UnsupportedPackageManagerError } = require('../errors');
const { isMultiProjectScan } = require('../is-multi-project-scan');
const { hasFeatureFlag } = require('../feature-flags');
const {
  PNPM_FEATURE_FLAG,
  DOTNET_WITHOUT_PUBLISH_FEATURE_FLAG,
  MAVEN_DVERBOSE_EXHAUSTIVE_DEPS_FF,
} = require('../package-managers');

async function test(root, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }
  if (!options) {
    options = {};
  }
  const promise = executeTest(root, options);
  if (callback) {
    promise
      .then((res) => {
        callback(null, res);
      })
      .catch(callback);
  }
  return promise;
}

async function executeTest(root, options) {
  let hasPnpmSupport = false;
  let hasImprovedDotnetWithoutPublish = false;
  let enableMavenDverboseExhaustiveDeps = false;
  try {
    hasPnpmSupport = await hasFeatureFlag(PNPM_FEATURE_FLAG, options);
    if (options['dotnet-runtime-resolution']) {
      hasImprovedDotnetWithoutPublish = await hasFeatureFlag(
        DOTNET_WITHOUT_PUBLISH_FEATURE_FLAG,
        options,
      );
      if (hasImprovedDotnetWithoutPublish) {
        options.useImprovedDotnetWithoutPublish = true;
      }
    }
  } catch (err) {
    hasPnpmSupport = false;
  }

  try {
    const args = options['_doubleDashArgs'] || [];
    const verboseEnabled =
      args.includes('-Dverbose') ||
      args.includes('-Dverbose=true') ||
      !!options['print-graph'];
    if (verboseEnabled) {
      enableMavenDverboseExhaustiveDeps = await hasFeatureFlag(
        MAVEN_DVERBOSE_EXHAUSTIVE_DEPS_FF,
        options,
      );
      if (enableMavenDverboseExhaustiveDeps) {
        options.mavenVerboseIncludeAllVersions =
          enableMavenDverboseExhaustiveDeps;
      }
    }
  } catch (err) {
    enableMavenDverboseExhaustiveDeps = false;
  }

  try {
    const featureFlags = hasPnpmSupport
      ? new Set([PNPM_FEATURE_FLAG])
      : new Set([]);

    if (!options.allProjects) {
      options.packageManager = detect.detectPackageManager(
        root,
        options,
        featureFlags,
      );
    }

    return run(root, options, featureFlags).then((results) => {
      for (const res of results) {
        if (!res.packageManager) {
          res.packageManager = options.packageManager;
        }
      }
      if (results.length === 1) {
        // Return only one result if only one found as this is the default usecase
        return results[0];
      }
      // For gradle, yarnWorkspaces, allProjects we may be returning more than one result
      return results;
    });
  } catch (error) {
    return Promise.reject(
      chalk.red.bold(error.message ? error.message : error),
    );
  }
}

function run(root, options, featureFlags) {
  const projectType = options.packageManager;
  validateProjectType(options, projectType, featureFlags);
  return runTest(projectType, root, options, featureFlags);
}

function validateProjectType(options, projectType, featureFlags) {
  if (projectType === 'pnpm' && !featureFlags.has(PNPM_FEATURE_FLAG)) {
    throw new UnsupportedPackageManagerError(projectType);
  }

  if (
    !(
      options.docker ||
      isMultiProjectScan(options) ||
      pm.SUPPORTED_PACKAGE_MANAGER_NAME[projectType]
    )
  ) {
    throw new UnsupportedPackageManagerError(projectType);
  }
}
