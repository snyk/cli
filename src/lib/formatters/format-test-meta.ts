import chalk from 'chalk';
import { rightPadWithSpaces } from '../../lib/right-pad';
import { TestOptions, Options } from '../../lib/types';
import { TestResult } from '../../lib/snyk-test/legacy';
import { IacTestResponse } from '../../lib/snyk-test/iac-test-result';

export function getOSFromTestResult(
  res: TestResult,
): { prettyName?: string } | null {
  // Check if we have scanResult with facts
  if (!res.scanResult?.facts) {
    return null;
  }

  const osReleaseFact = res.scanResult.facts.find(
    (fact) => fact.type === 'imageOsReleasePrettyName',
  );

  if (osReleaseFact?.data) {
    return {
      prettyName: osReleaseFact.data,
    };
  }

  return null;
}

export function formatTestMeta(
  res: TestResult | IacTestResponse,
  options: Options & TestOptions,
): string {
  const padToLength = 19; // chars to align
  const packageManager = res.packageManager || options.packageManager;
  const targetFile = res.targetFile || res.displayTargetFile || options.file;
  const openSource = res.isPrivate ? 'no' : 'yes';
  const meta = res.org
    ? [chalk.bold(rightPadWithSpaces('Organization: ', padToLength)) + res.org]
    : [];
  meta.push(
    chalk.bold(rightPadWithSpaces('Package manager: ', padToLength)) +
      packageManager,
  );
  if (targetFile) {
    meta.push(
      chalk.bold(rightPadWithSpaces('Target file: ', padToLength)) + targetFile,
    );
  }
  if (res.projectName) {
    meta.push(
      chalk.bold(rightPadWithSpaces('Project name: ', padToLength)) +
        res.projectName,
    );
  }
  if (options.docker) {
    meta.push(
      chalk.bold(rightPadWithSpaces('Docker image: ', padToLength)) +
        options.path,
    );

    if (res.platform) {
      meta.push(
        chalk.bold(rightPadWithSpaces('Platform: ', padToLength)) +
          res.platform,
      );
    }

    const legacyRes: TestResult = res as TestResult;
    const osInfo = getOSFromTestResult(legacyRes);
    if (osInfo?.prettyName) {
      meta.push(
        chalk.bold(rightPadWithSpaces('Target OS: ', padToLength)) +
          osInfo.prettyName,
      );
    }
  } else {
    meta.push(
      chalk.bold(rightPadWithSpaces('Open source: ', padToLength)) + openSource,
    );
    meta.push(
      chalk.bold(rightPadWithSpaces('Project path: ', padToLength)) +
        options.path,
    );
  }
  if (res.payloadType !== 'k8sconfig') {
    const legacyRes: TestResult = res as TestResult;
    if (legacyRes.docker && legacyRes.docker.baseImage) {
      meta.push(
        chalk.bold(rightPadWithSpaces('Base image: ', padToLength)) +
          legacyRes.docker.baseImage,
      );
    }

    if (legacyRes.filesystemPolicy) {
      meta.push(
        chalk.bold(rightPadWithSpaces('Local Snyk policy: ', padToLength)) +
          chalk.green('found'),
      );
      if (
        legacyRes.ignoreSettings &&
        legacyRes.ignoreSettings.disregardFilesystemIgnores
      ) {
        meta.push(
          chalk.bold(
            rightPadWithSpaces('Local Snyk policy ignored: ', padToLength),
          ) + chalk.red('yes'),
        );
      }
    }
    if (legacyRes.licensesPolicy) {
      meta.push(
        chalk.bold(rightPadWithSpaces('Licenses: ', padToLength)) +
          chalk.green('enabled'),
      );
    }
  }

  return meta.join('\n');
}
