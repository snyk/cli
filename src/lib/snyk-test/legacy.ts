import * as values from 'lodash.values';
import * as depGraphLib from '@snyk/dep-graph';
import {
  DepsFilePaths,
  ScanResult,
  FileSignaturesDetails,
} from '../ecosystems/types';
import { SupportedPackageManagers } from '../package-managers';
import { Options, SupportedProjectTypes, TestOptions } from '../types';
import { SEVERITIES } from './common';
import { AppliedPolicyRules } from '../formatters/types';

interface Pkg {
  name: string;
  version?: string;
}

export interface Patch {
  version: string;
  id: string;
  urls: string[];
  modificationTime: string;
}

export enum SEVERITY {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export interface VulnMetaData {
  id: string;
  title: string;
  description: string;
  type: 'license' | 'vuln';
  name: string;
  info: string;
  severity: SEVERITY;
  severityValue: number;
  isNew: boolean;
  version: string;
  packageManager: SupportedPackageManagers | 'upstream';
}

export interface GroupedVuln {
  list: AnnotatedIssue[];
  metadata: VulnMetaData;
  isIgnored: boolean;
  title: string;
  note: string | false;
  severity: SEVERITY;
  originalSeverity?: SEVERITY;
  isNew: boolean;
  name: string;
  version: string;
  isFixable: boolean;
  fixedIn: string[];
  legalInstructionsArray?: LegalInstruction[];
  appliedPolicyRules?: AppliedPolicyRules;
}

export interface LegalInstruction {
  licenseName: string;
  legalContent: string;
}
export interface IssueData {
  id: string;
  packageName: string;
  version: string;
  moduleName?: string;
  below: string; // Vulnerable below version
  semver: {
    vulnerable: string | string[];
    vulnerableHashes?: string[];
    vulnerableByDistro?: {
      [distroNameAndVersion: string]: string[];
    };
  };
  patches: Patch[];
  isNew: boolean;
  description: string;
  title: string;
  severity: SEVERITY;
  fixedIn: string[];
  legalInstructions?: string;
  packageManager?: SupportedProjectTypes;
  from?: string[];
  name?: string;
}

export interface IssueDataUnmanaged extends IssueData {
  upgradePath?: (string | boolean)[];
  isPatchable?: boolean;
}

export type CallPath = string[];

interface AnnotatedIssue extends IssueData {
  credit: string[];
  name: string;
  version: string;
  from: string[];
  upgradePath: Array<string | boolean>;
  isUpgradable: boolean;
  isPatchable: boolean;
  severity: SEVERITY;
  originalSeverity?: SEVERITY;
  cvssScore?: number;
  lineNumber?: number;

  // These fields present for "node_module" based scans to allow remediation
  bundled?: any;
  shrinkwrap?: any;
  __filename?: string;
  parentDepType: string;

  type?: 'license';
  title: string;
  patch?: any;
  note?: string | false;
  publicationTime?: string;

  identifiers?: {
    [name: string]: string[];
  };
}

// Mixin, to be added to GroupedVuln / AnnotatedIssue
export interface DockerIssue {
  nearestFixedInVersion?: string;
  dockerfileInstruction?: any;
  dockerBaseImage?: any;
}

export interface IgnoreSettings {
  adminOnly: boolean;
  reasonRequired: boolean;
  disregardFilesystemIgnores: boolean;
}

export interface BasicResultData {
  ok: boolean;
  payloadType?: string;
  org: string;
  isPrivate: boolean;
  summary: string;
  packageManager?: SupportedProjectTypes;
  severityThreshold?: string;
  platform?: string;
}

export interface LegacyVulnApiResult extends BasicResultData {
  vulnerabilities: AnnotatedIssue[];
  dependencyCount: number;
  policy: string;
  licensesPolicy: object | null;
  ignoreSettings: IgnoreSettings | null;
  docker?: {
    baseImage?: any;
    binariesVulns?: unknown;
    baseImageRemediation?: BaseImageRemediation;
  };
  projectId?: string;
  filesystemPolicy?: boolean;
  uniqueCount?: any;
  remediation?: RemediationChanges;
}

export interface BaseImageRemediation {
  code: string;
  advice: BaseImageRemediationAdvice[];
  message?: string; // TODO: check if this is still being sent
}

export interface BaseImageRemediationAdvice {
  message: string;
  bold?: boolean;
  color?: string;
}

export interface TestResult extends LegacyVulnApiResult {
  targetFile?: string;
  projectName?: string;
  targetFilePath?: string;
  displayTargetFile?: string; // used for display only
  foundProjectCount?: number;
  scanResult?: ScanResult;
  hasUnknownVersions?: boolean;
  path?: string;
}

interface UpgradePathItem {
  name: string;
  version: string;
  newVersion?: string;
  isDropped?: boolean;
}

interface UpgradePath {
  path: UpgradePathItem[];
}

interface FixInfo {
  upgradePaths: UpgradePath[];
  isPatchable: boolean;
  nearestFixedInVersion?: string;
}

export interface AffectedPackages {
  [pkgId: string]: {
    pkg: Pkg;
    issues: {
      [issueId: string]: Issue;
    };
  };
}

interface TestDepGraphResult {
  issuesData: {
    [issueId: string]: IssueData;
  };
  affectedPkgs: AffectedPackages;
  docker: {
    binariesVulns?: TestDepGraphResult;
    baseImage?: any;
  };
  remediation?: RemediationChanges;
}

export interface Issue {
  pkgName: string;
  pkgVersion?: string;
  issueId: string;
  fixInfo: FixInfo;
}

export interface TestDependenciesResult {
  issuesData: {
    [issueId: string]: IssueDataUnmanaged;
  };
  issues: Issue[];
  docker?: {
    baseImage: string;
    baseImageRemediation: BaseImageRemediation;
    binariesVulns: TestDepGraphResult;
  };
  remediation?: RemediationChanges;
  depsFilePaths?: DepsFilePaths;
  depGraphData: depGraphLib.DepGraphData;
  fileSignaturesDetails: FileSignaturesDetails;
  vulnerabilities: IssueData[];
  path: string;
  dependencyCount: number;
  packageManager: SupportedProjectTypes;
}

export interface TestDepGraphMeta {
  isPublic: boolean;
  isLicensesEnabled: boolean;
  licensesPolicy?: {
    severities: {
      [type: string]: string;
    };
  };
  projectId?: string;
  ignoreSettings?: IgnoreSettings;
  policy: string;
  org: string;
}

export interface TestDepGraphResponse {
  result: TestDepGraphResult;
  meta: TestDepGraphMeta;
}

export interface TestDependenciesResponse {
  result: TestDependenciesResult;
  meta: TestDepGraphMeta;
}

export interface Ignores {
  [path: string]: {
    paths: string[][];
    meta: {
      days?: number;
      reason?: string;
    };
  };
}

export interface PatchObject {
  [name: string]: {
    patched: string;
  };
}

export interface Upgrade {
  upgradeTo: string; // name@version
}

export interface UpgradeVulns extends Upgrade {
  vulns: string[];
}

export interface UpgradeRemediation extends UpgradeVulns {
  upgrades: string[];
}

export interface PatchRemediation {
  paths: PatchObject[];
}

export interface DependencyUpdates {
  [from: string]: UpgradeRemediation;
}

export interface PinRemediation extends UpgradeVulns {
  isTransitive: boolean;
}

export interface DependencyPins {
  [name: string]: PinRemediation;
}

// Remediation changes to be applied to the project,
// including information on all and unresolved issues.
export interface RemediationChanges {
  unresolved: IssueData[];
  upgrade: DependencyUpdates;
  patch: {
    [name: string]: PatchRemediation;
  };
  ignore: unknown;
  pin: DependencyPins;
}

function convertTestDepGraphResultToLegacy(
  res: TestDepGraphResponse,
  depGraph: depGraphLib.DepGraph,
  packageManager: SupportedProjectTypes | undefined,
  options: Options & TestOptions,
): LegacyVulnApiResult {
  const result = res.result;

  const upgradePathsMap = new Map<string, string[]>();

  for (const pkgInfo of values(result.affectedPkgs)) {
    for (const pkgIssue of values(pkgInfo.issues)) {
      if (pkgIssue.fixInfo && pkgIssue.fixInfo.upgradePaths) {
        for (const upgradePath of pkgIssue.fixInfo.upgradePaths) {
          const legacyFromPath = pkgPathToLegacyPath(upgradePath.path);
          const vulnPathString = getVulnPathString(
            pkgIssue.issueId,
            legacyFromPath,
          );
          upgradePathsMap[vulnPathString] = toLegacyUpgradePath(
            upgradePath.path,
          );
        }
      }
    }
  }

  // generate the legacy vulns array (vuln-data + metada per vulnerable path).
  //   use the upgradePathsMap to find available upgrade-paths
  const vulns: AnnotatedIssue[] = [];

  for (const pkgInfo of values(result.affectedPkgs)) {
    for (const vulnPkgPath of depGraph.pkgPathsToRoot(pkgInfo.pkg)) {
      const legacyFromPath = pkgPathToLegacyPath(vulnPkgPath.reverse());
      for (const pkgIssue of values(pkgInfo.issues)) {
        const vulnPathString = getVulnPathString(
          pkgIssue.issueId,
          legacyFromPath,
        );
        const upgradePath = upgradePathsMap[vulnPathString] || [];

        // TODO: we need the full issue-data for every path only for the --json output,
        //   consider picking only the required fields,
        //   and append the full data only for --json, to minimize chance of out-of-memory
        const annotatedIssue = Object.assign(
          {},
          result.issuesData[pkgIssue.issueId],
          {
            from: legacyFromPath,
            upgradePath,
            isUpgradable: !!upgradePath[0] || !!upgradePath[1],
            isPatchable: pkgIssue.fixInfo.isPatchable,
            name: pkgInfo.pkg.name,
            version: pkgInfo.pkg.version as string,
            nearestFixedInVersion: pkgIssue.fixInfo.nearestFixedInVersion,
          },
        ) as AnnotatedIssue & DockerIssue; // TODO(kyegupov): get rid of type assertion

        vulns.push(annotatedIssue);
      }
    }
  }

  const dockerRes = result.docker;

  if (dockerRes && dockerRes.binariesVulns) {
    const binariesVulns = dockerRes.binariesVulns;
    for (const pkgInfo of values(binariesVulns.affectedPkgs)) {
      for (const pkgIssue of values(pkgInfo.issues)) {
        const pkgAndVersion = (pkgInfo.pkg.name +
          '@' +
          pkgInfo.pkg.version) as string;
        const annotatedIssue = Object.assign(
          {},
          binariesVulns.issuesData[pkgIssue.issueId],
          {
            from: ['Upstream', pkgAndVersion],
            upgradePath: [],
            isUpgradable: false,
            isPatchable: false,
            name: pkgInfo.pkg.name,
            version: pkgInfo.pkg.version as string,
            nearestFixedInVersion: pkgIssue.fixInfo.nearestFixedInVersion,
          },
        ) as any as AnnotatedIssue; // TODO(kyegupov): get rid of forced type assertion
        vulns.push(annotatedIssue);
      }
    }
  }

  const meta = res.meta || {};

  const severityThreshold =
    options.severityThreshold === SEVERITY.LOW
      ? undefined
      : options.severityThreshold;

  const legacyRes: LegacyVulnApiResult = {
    vulnerabilities: vulns,
    ok: vulns.length === 0,
    dependencyCount: depGraph.getPkgs().length - 1,
    org: meta.org,
    policy: meta.policy,
    isPrivate: !meta.isPublic,
    licensesPolicy: meta.licensesPolicy || null,
    packageManager,
    projectId: meta.projectId,
    ignoreSettings: meta.ignoreSettings || null,
    docker: result.docker,
    summary: getSummary(vulns, severityThreshold),
    severityThreshold,
    remediation: result.remediation,
  };

  return legacyRes;
}

function getVulnPathString(issueId: string, vulnPath: string[]) {
  return issueId + '|' + JSON.stringify(vulnPath);
}

function pkgPathToLegacyPath(pkgPath: Pkg[]): string[] {
  return pkgPath.map(toLegacyPkgId);
}

function toLegacyUpgradePath(
  upgradePath: UpgradePathItem[],
): Array<string | boolean> {
  return upgradePath
    .filter((item) => !item.isDropped)
    .map((item) => {
      if (!item.newVersion) {
        return false;
      }

      return `${item.name}@${item.newVersion}`;
    });
}

function toLegacyPkgId(pkg: Pkg) {
  return `${pkg.name}@${pkg.version || '*'}`;
}

function getSummary(vulns: object[], severityThreshold?: SEVERITY): string {
  const count = vulns.length;
  let countText = '' + count;
  const severityFilters: string[] = [];
  const severitiesArray = SEVERITIES.map((s) => s.verboseName);
  if (severityThreshold) {
    severitiesArray
      .slice(severitiesArray.indexOf(severityThreshold))
      .forEach((sev) => {
        severityFilters.push(sev);
      });
  }

  if (!count) {
    if (severityFilters.length) {
      return `No ${severityFilters.join(' or ')} severity vulnerabilities`;
    }
    return 'No known vulnerabilities';
  }

  if (severityFilters.length) {
    countText += ' ' + severityFilters.join(' or ') + ' severity';
  }

  return `${countText} vulnerable dependency ${pl('path', count)}`;
}

function pl(word, count) {
  const ext = {
    y: 'ies',
    default: 's',
  };

  const last = word.split('').pop();

  if (count > 1) {
    return word.slice(0, -1) + (ext[last] || last + ext.default);
  }

  return word;
}

export { convertTestDepGraphResultToLegacy, AnnotatedIssue };
