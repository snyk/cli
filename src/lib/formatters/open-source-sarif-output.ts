import * as sarif from 'sarif';
import * as upperFirst from 'lodash.upperfirst';
import * as groupBy from 'lodash.groupby';
import * as map from 'lodash.map';

import { TestResult, AnnotatedIssue } from '../snyk-test/legacy';
import { getResults } from './get-sarif-result';
import { getVersion } from '../../lib/version';

const LOCK_FILES_TO_MANIFEST_MAP = {
  'Gemfile.lock': 'Gemfile',
  'package-lock.json': 'package.json',
  'yarn.lock': 'package.json',
  'pnpm-lock.yaml': 'package.json',
  'Gopkg.lock': 'Gopkg.toml',
  'go.sum': 'go.mod',
  'composer.lock': 'composer.json',
  'Podfile.lock': 'Podfile',
  'poetry.lock': 'pyproject.toml',
};

export function createSarifOutputForOpenSource(
  testResults: TestResult[],
): sarif.Log {
  return {
    $schema:
      'https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json',
    version: '2.1.0',
    runs: testResults.map(replaceLockfileWithManifest).map((testResult) => ({
      automationDetails : getAutomationDetails(testResult),
      tool: {
        driver: {
          name: 'Snyk Open Source',
          semanticVersion: getVersion(),
          version: getVersion(),
          informationUri: 'https://docs.snyk.io/',
          properties: {
            artifactsScanned: testResult.dependencyCount,
          },
          rules: getRules(testResult),
        },
      },
      results: getResults(testResult),
    })),
  };
}

// Github anncouned changes to their SARIF upload -- https://github.blog/changelog/2024-05-06-code-scanning-will-stop-combining-runs-from-a-single-upload/
// the impact is when a SARIF that is being uploaded, each run must have unique category, as defined by GitHub here, https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#runautomationdetails-object
// This presents a new problem of when a file is removed from source since GH will not have an empty result to close any previously opened items since GH open/closes
// based on the SARIF tool.driver.name + Category. Open source's solution is the most obvious, inlcude the targetFile. Snyk-iac, is using this field set to a static "snyk-iac". Combing what
// was being done there with the file name to generate the unique value. Using | as a separator to make it easier to parse out tool vs targetFile. 
function getAutomationDetails(testResult: TestResult)
{
  let automationId = !!process.env.SET_AUTOMATION_DETAILS_ID ? `snyk-sca|${testResult.displayTargetFile || testResult.targetFile }/` : ""
  return {
    id : automationId,
  };
}

function replaceLockfileWithManifest(testResult: TestResult): TestResult {
  let targetFile = testResult.displayTargetFile || '';
  for (const [key, replacer] of Object.entries(LOCK_FILES_TO_MANIFEST_MAP)) {
    targetFile = targetFile.replace(new RegExp(key, 'g'), replacer);
  }
  return {
    ...testResult,
    vulnerabilities: testResult.vulnerabilities || [],
    displayTargetFile: targetFile,
  };
}

export function getRules(testResult: TestResult): sarif.ReportingDescriptor[] {
  const groupedVulnerabilities = groupBy(testResult.vulnerabilities, 'id');
  return map(
    groupedVulnerabilities,
    ([vuln, ...moreVulns]: AnnotatedIssue[]): sarif.ReportingDescriptor => {
      const cves = vuln.identifiers?.CVE?.join();
      return {
        id: vuln.id,
        shortDescription: {
          text: `${upperFirst(vuln.severity)} severity - ${
            vuln.title
          } vulnerability in ${vuln.packageName}`,
        },
        fullDescription: {
          text: cves
            ? `(${cves}) ${vuln.name}@${vuln.version}`
            : `${vuln.name}@${vuln.version}`,
        },
        help: {
          text: '',
          markdown: `* Package Manager: ${testResult.packageManager}
* ${vuln.type === 'license' ? 'Module' : 'Vulnerable module'}: ${vuln.name}
* Introduced through: ${getIntroducedThrough(vuln)}
#### Detailed paths
${[vuln, ...moreVulns]
  .map((item) => `* _Introduced through_: ${item.from.join(' â€º ')}`)
  .join('\n')}
${vuln.description}`.replace(/##\s/g, '# '),
        },
        properties: {
          tags: [
            'security',
            ...(vuln.identifiers?.CWE || []),
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            testResult.packageManager!,
          ],
          cvssv3_baseScore: vuln.cvssScore, // AWS
          'security-severity': String(vuln.cvssScore), // GitHub
        },
      };
    },
  );
}

function getIntroducedThrough(vuln: AnnotatedIssue) {
  const [firstFrom, secondFrom] = vuln.from || [];

  return vuln.from.length > 2
    ? `${firstFrom}, ${secondFrom} and others`
    : vuln.from.length === 2
    ? `${firstFrom} and ${secondFrom}`
    : firstFrom;
}
