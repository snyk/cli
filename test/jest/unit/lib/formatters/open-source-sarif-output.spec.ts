import { createSarifOutputForOpenSource } from '../../../../../src/lib/formatters/open-source-sarif-output';
import { SEVERITY, TestResult } from '../../../../../src/lib/snyk-test/legacy';

describe('createSarifOutputForOpenSource', () => {
  it('general', () => {
    const testFile = getTestResult(
      {}, // testResultOverride
      {
        cvssScore: 7.5, // vulnOverride
      },
    );
    const sarif = createSarifOutputForOpenSource([testFile]);
    expect(sarif).toMatchSnapshot();
  });

  describe('replace lock-file to manifest-file', () => {
    const lockFiles = [
      'Gemfile.lock',
      'package-lock.json',
      'yarn.lock',
      'Gopkg.lock',
      'go.sum',
      'composer.lock',
      'Podfile.lock',
      'poetry.lock',
    ];

    lockFiles.forEach((lockFileName) =>
      it(lockFileName, () => {
        const time = Date.now();
        const testFile = getTestResult({
          displayTargetFile: `${time}/${lockFileName}`,
        });
        const sarif = createSarifOutputForOpenSource([testFile]);
        const uri =
          sarif.runs?.[0]?.results?.[0].locations?.[0]?.physicalLocation?.artifactLocation?.uri?.replace(
            `${time}/`,
            '',
          );
        expect(uri).toMatchSnapshot();
      }),
    );
  });
});

function getTestResult(testResultOverride = {}, vulnOverride = {}): TestResult {
  return {
    vulnerabilities: [
      {
        below: '',
        credit: ['Unknown'],
        description: '## Overview\n',
        fixedIn: ['6.12.3'],
        id: 'SNYK-JS-AJV-584908',
        identifiers: {
          CVE: ['CVE-2020-15366'],
          CWE: ['CWE-400'],
        },
        moduleName: 'ajv',
        packageManager: 'npm',
        packageName: 'ajv',
        patches: [],
        publicationTime: '2020-07-16T13:58:04Z',
        semver: {
          vulnerable: ['<6.12.3'],
        },
        severity: SEVERITY.CRITICAL,
        title: 'Prototype Pollution',
        from: [
          'PROJECT_NAME@1.0.0',
          'jimp@0.2.28',
          'request@2.88.2',
          'har-validator@5.1.3',
          'ajv@6.12.2',
        ],
        upgradePath: [
          false,
          'jimp@0.2.28',
          'request@2.88.2',
          'har-validator@5.1.3',
          'ajv@6.12.3',
        ],
        isUpgradable: true,
        isPatchable: false,
        name: 'ajv',
        version: '6.12.2',
        __filename: 'node_modules/ajv/package.json',
        parentDepType: 'prod',
        isNew: false,
        ...vulnOverride,
      },
    ],
    ok: false,
    dependencyCount: 969,
    org: 'ORG',
    policy: '',
    isPrivate: true,
    licensesPolicy: {
      severities: {},
      orgLicenseRules: {},
    },
    packageManager: 'npm',
    ignoreSettings: null,
    summary: '165 vulnerable dependency paths',
    filesystemPolicy: false,
    uniqueCount: 42,
    projectName: 'PROJECT_NAME',
    foundProjectCount: 22,
    displayTargetFile: 'package.json',
    ...testResultOverride,
  };
}
