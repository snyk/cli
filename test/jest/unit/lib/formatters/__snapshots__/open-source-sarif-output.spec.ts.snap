// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSarifOutputForOpenSource general 1`] = `
{
  "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
  "runs": [
    {
      "automationDetails": {
        "id": "",
      },
      "results": [
        {
          "fixes": [
            {
              "artifactChanges": [
                {
                  "artifactLocation": {
                    "uri": "package.json",
                  },
                  "replacements": [
                    {
                      "deletedRegion": {
                        "startLine": 1,
                      },
                      "insertedContent": {
                        "text": "jimp@0.2.28",
                      },
                    },
                  ],
                },
              ],
              "description": {
                "text": "Upgrade to jimp@0.2.28",
              },
            },
          ],
          "level": "error",
          "locations": [
            {
              "logicalLocations": [
                {
                  "fullyQualifiedName": "ajv@6.12.2",
                },
              ],
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "package.json",
                },
                "region": {
                  "startLine": 1,
                },
              },
            },
          ],
          "message": {
            "text": "This file introduces a vulnerable ajv package with a critical severity vulnerability.",
          },
          "ruleId": "SNYK-JS-AJV-584908",
        },
      ],
      "tool": {
        "driver": {
          "informationUri": "https://docs.snyk.io/",
          "name": "Snyk Open Source",
          "properties": {
            "artifactsScanned": 969,
          },
          "rules": [
            {
              "fullDescription": {
                "text": "(CVE-2020-15366) ajv@6.12.2",
              },
              "help": {
                "markdown": "* Package Manager: npm
* Vulnerable module: ajv
* Introduced through: PROJECT_NAME@1.0.0, jimp@0.2.28 and others
### Detailed paths
* _Introduced through_: PROJECT_NAME@1.0.0 › jimp@0.2.28 › request@2.88.2 › har-validator@5.1.3 › ajv@6.12.2
# Overview
",
                "text": "",
              },
              "id": "SNYK-JS-AJV-584908",
              "properties": {
                "cvssv3_baseScore": 7.5,
                "security-severity": "7.5",
                "tags": [
                  "security",
                  "CWE-400",
                  "npm",
                ],
              },
              "shortDescription": {
                "text": "Critical severity - Prototype Pollution vulnerability in ajv",
              },
            },
          ],
          "semanticVersion": "1.0.0-monorepo",
          "version": "1.0.0-monorepo",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file Gemfile.lock 1`] = `"Gemfile"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file Gopkg.lock 1`] = `"Gopkg.toml"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file Podfile.lock 1`] = `"Podfile"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file composer.lock 1`] = `"composer.json"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file go.sum 1`] = `"go.mod"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file package-lock.json 1`] = `"package.json"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file poetry.lock 1`] = `"pyproject.toml"`;

exports[`createSarifOutputForOpenSource replace lock-file to manifest-file yarn.lock 1`] = `"package.json"`;
